# Streamlit-HGB

A streamlit component example of a hybrid genome browser

![streamlit](Streamlit.png)

* virtualenv

[hgb](https://github.com/6br/hgb) 

```bash
$ python3 -m venv venv  # create venv
$ . venv/bin/activate   # activate venv
$ pip install streamlit_hgb pandas gffutils argh # install streamlit and dependencies
```

And then put BAM and BAI files on this foilder.

```bash
export STREAMLIT_HOST=$(hostname -i)
export HGB_BIN=/*location_to_hgb_binary*/

streamlit run hgb.py 
```

HGB_BIN is usually `/dir/to/cloned/hgb/target/release/hgb`. Specify HGB_BIN, otherwise set the path to hgb as `export PATH=$PATH:/dir/to/hgb/`.

## Example of `config.yaml`

There is two ways to modify the input file of the browser.

1. Directly modify `hgb.py`
2. Write `config.yaml` at the same directory of `hgb.py`

```yaml
hg38:
  range: 
    - chr1:1-1000001
  default:
    - "/path/to/bamfile1.bam"
  alignments: 
    - "/path/to/bamfile1.bam"
    - "/path/to/bamfile2.bam"
  db:
    - "/path/to/gff.db"
```

* Range: a default range to visualize
* Default: a default set of BAM file to be loaded
* Alignments: a list of BAM file
* DB: a gff database generated by gffutils

All input bam files must be indexed by `samtools index` or compatibles.

## Generate GFF database

```bash
$ pip install gffutils
$ ./gffutils-cli create --disable-infer-genes input.gtf
```

